### PR TITLE
[hwasan] Store some report data early

### PR DESCRIPTION
**Pull request by vitalybuka**
Please review them one by one in order, and let me know which are OK.
It's mostly code shuffling.

The goal is to make HWASA collect all needed info the first, and printout later.
This way we avoid the printing changes HWASAN metadata.

### CODE REVIEW

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 1]
```diff
@@ -37,7 +37,7 @@ namespace __hwasan {
 
 class ScopedReport {
  public:
-  ScopedReport(bool fatal = false) : fatal(fatal) {
+  ScopedReport(bool fatal) : fatal(fatal) {
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:40]: Unrelated change, right?

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 2]
```diff
@@ -373,7 +375,7 @@ static void ShowHeapOrGlobalCandidate(uptr untagged_addr, tag_t *candidate,
   }
 }
 
-void PrintAddressDescription(
+static void PrintAddressDescription(
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp]: Why don't we pull all the static into the anon namespace

Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp]: There is some LLVM rule to prefer static over {}
but it will be a method anyway

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 3]
```diff
@@ -608,7 +610,7 @@ static void PrintAddressDescription(
 class InvalidFreeReport : public BaseReport {
  public:
   InvalidFreeReport(StackTrace *stack, uptr tagged_addr)
-      : BaseReport(stack, flags()->halt_on_error, tagged_addr) {}
+      : BaseReport(stack, flags()->halt_on_error, tagged_addr, 0) {}
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:634]: optional: make BaseReport::kNoAccessSize = 0, and use that here?

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 2 above)
Comment (fmayer) [reply compiler-rt/lib/hwasan/hwasan_report.cpp]: For posterity: https://llvm.org/docs/CodingStandards.html#anonymous-namespaces

> Because of this, we have a simple guideline: make anonymous namespaces as small as possible, and only use them for class declarations. For example:

**Comment by fmayer** [comment]
LGTM

**Comment by vitalybuka** [comment]
@fmayer Please continue with these:

624793f88d063ae9fbdc72028bef6c01ee427198 Collect heap allocations early (#66682)
bcbbdb5ea482e1d2a47b35532cfbdd6acef0a45b [NFC][hwasan] Extract announce_by_id (#66682)
ac525a4f593439972cbce4bc559103d29eb59e97 [NFC][hwasan] Extract a few BaseReport::Copy methods (#66682)
9ac8957f87c4ee11a66ba3cd9fcb7b6b5e7df17b [NFC][hwasan] Find overflow candidate early (#66682)
da60da5dfb9e1ee2ae5ef515d067acc478edd375 [NFC][hwasan] Make ShowHeapOrGlobalCandidate a method (#66682)

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 4]
```diff
@@ -427,74 +429,128 @@ class BaseReport {
...
+      candidate_tag_ptr = right;
+      break;
+    }
+    ++right;
+  }
+
+  constexpr auto kCloseCandidateDistance = 1;
+  candidate.is_close = candidate_distance <= kCloseCandidateDistance;
+
+  {
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:506]: What is this scope for? The scope seems to end anyway at the end of the function?

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 5]
```diff
@@ -427,74 +429,128 @@ class BaseReport {
...
+    candidate.after = candidate_tag_ptr == left;
+    candidate.untagged_addr =
+        ShadowToMem(reinterpret_cast<uptr>(candidate_tag_ptr));
+    HwasanChunkView chunk = FindHeapChunkByAddress(candidate.untagged_addr);
+    if (chunk.IsAllocated()) {
+      candidate.heap.is_allocated = chunk.IsAllocated();
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:512]: This is always true

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 6]
```diff
@@ -404,12 +404,14 @@ class BaseReport {
         stack_allocations[stack_allocations_count++].CopyFrom(t);
       }
     });
+
+    FindBufferOverflowCandidate();
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:395]: I think this would be clearer if `FindBufferOverflowCandidate` returned the struct, and this would read `candidate = FindBufferOverflowCandidate()`.

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 7]
```diff
@@ -585,6 +585,13 @@ void BaseReport::PrintAddressDescription() const {
         untagged_addr - heap.begin, d.Default());
   }
 
+  auto announce_by_id = [](u32 thread_id) {
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:613]: I am not a fan, I don't think this indirection adds anything.

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 8]
```diff
@@ -614,42 +639,39 @@ void BaseReport::PrintAddressDescription() const {
...
-      Printf("hwasan_dev_note_num_matching_addrs: %zd\n", num_matching_addrs);
-      Printf("hwasan_dev_note_num_matching_addrs_4b: %zd\n",
-             num_matching_addrs_4b);
-
-      t->Announce();
-      num_descriptions_printed++;
-    }
-  });
+  for (uptr i = 0; i < heap_allocations_count; ++i) {
+    auto &ha = heap_allocations[i];
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:643]: I think this can be `const`

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 4 above)
Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp:506]: Yes. it's redundunt, removed.

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 7 above)
Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp:669]: we used it twice
so either repeat or indirection

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 6 above)
Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp]: done

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 5 above)
Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp]: done

**Comment by vitalybuka** [comment]
@fmayer PTAL

61289c01fa2972bb842c08706316e9fa1c898041 [NFC][hwasan] Stored shadow bytes early (#66682)
bfee0f767efbde09553ee0a910077d0eb4d19825 [NFC][hwasan] Stored tail early (#66682)
29e4f2bf21185e052625d81d1293cdd2c56a99af [NFC][hwasan] Use stored chunk in TailOverwrittenReport (#66682)
40ac72793accfdee3052567a5953bb3233e8539e [NFC][hwasan] Set more fields by value and make them const (#66682)
03d8889dd852de5154eea58a6f3949b00cd2b2bd [NFC][hwasan] More consts in BaseReport (#66682)

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 9]
```diff
@@ -321,55 +322,61 @@ static uptr GetGlobalSizeFromDescriptor(uptr ptr) {
...
-  tag_t *end_row = center_row_beg + row_len * ((num_rows + 1) / 2);
-  InternalScopedString s;
-  for (tag_t *row = beg_row; row < end_row; row += row_len) {
+  uptr center_row_beg = RoundDownTo(addr, row_len);
+  uptr beg_row = center_row_beg - row_len * (num_rows / 2);
+  uptr end_row = center_row_beg + row_len * ((num_rows + 1) / 2);
+  for (uptr row = beg_row; row < end_row; row += row_len) {
     s.Append(row == center_row_beg ? "=>" : "  ");
-    s.AppendF("%p:", (void *)ShadowToMem(reinterpret_cast<uptr>(row)));
+    s.AppendF("%p:", (void *)ShadowToMem(row));
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:335]: Drive by: can we get rid of the cast and use `%#lx`?

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 10]
```diff
@@ -427,6 +436,9 @@ class BaseReport {
     bool is_allocated = false;
   };
 
+  void CopyShadow();
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:439]: Should we do the same as with `CopyAllocations`, i.e. return and assign?

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 11]
```diff
@@ -446,8 +458,51 @@ class BaseReport {
   const HeapChunk heap;
   const Allocations allocations;
   const OverflowCandidate candidate;
+
+  struct {
+    uptr addr = 0;
+    tag_t tags[512] = {};
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:464]: should this be odd so we have the same amount of tags on both sides of the address?

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 12]
```diff
@@ -446,8 +458,51 @@ class BaseReport {
...
+    uptr granule_addr = ShadowToMem(tag_addr);
+    if (1 <= shadow.tags[i] && shadow.tags[i] <= kShadowAlignment &&
+        IsAccessibleMemoryRange(granule_addr, kShadowAlignment)) {
+      shadow.short_tags[i] =
+          *reinterpret_cast<tag_t *>(granule_addr + kShadowAlignment - 1);
+    }
+  }
+}
+
+tag_t BaseReport::GetTagCopy(uptr addr) const {
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:488]: Why do we expect for those to be called out of range? Should the checks be assert instead? `0` tag is not the same as "we don't know" really.

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 11 above)
Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp:464]: we will align on 16, so we want the same number of lines before and after (17lines x 16bytes)
so logic will be more complicated, but possible.
However if we want to keep printing outside (we may want a print tags function outside of reporting), then I'd rather go with larger snapshot, but simpler logic.
 
Another issue is that when we do a snapshot we don't use a pierce address. in one of usecases we add access size as offset.

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 10 above)
Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp:439]: Isn't that here already https://github.com/llvm/llvm-project/commit/40ac72793accfdee3052567a5953bb3233e8539e ?

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 9 above)
Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp:335]: even if it looks the same, I would prefer to use %p for pointers as intended.

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 10 above)
Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_report.cpp:439]: done

**Review by fmayer — APPROVED** [review]
